### PR TITLE
upgrade lifecycle handler

### PR DESCRIPTION
**Pull request by il3ven**
This is a proposal for an upgrade in lifecycle handler. It solves two main problems.
- We now know when an extractor has completed and we can then call it's transformer. This should solve problems like #154. I tested it and it does solve. The number of lines in soundxyz-metadata-transformation is the same as soundxyz-metadata-extraction.
- We now also know when a strategy has ended. This should solve problems like #146.

Additional benefit:
- We can now define a [path](https://github.com/neume-network/strategies/blob/2bff5f4fda838239de8426806d330eee79c85916/src/lifecycle.mjs#L202) for the crawler to take in advance.

@TimDaub Have a look at the implementation. I tested it with minimal data and it works. I couldn't test it completely because of no IPFS gateway access right now. I'll try to test more and iron out the bugs.

### CODE REVIEW

**Comment by TimDaub** [comment]
@reimertz can we set up @il3ven with an IPFS gateway access?

File: src/lifecycle.mjs [hunk 1]
```diff
@@ -112,111 +106,173 @@ async function transform(strategy, name, type) {
...
+  });
 
-    if (
-      messages.lifecycle &&
-      messages.lifecycle.length >= 1 &&
-      messages.lifecycle[0].type === "transformation"
-    ) {
-      log(`Ending extractor strategy with name "${strategy.module.name}"`);
-    }
+  const graph = [
```
Comment (TimDaub) [inline src/lifecycle.mjs:202]: this is amazing, I wanted to have something like this since a while!

**Comment by TimDaub** [comment]
this may also resolve: https://github.com/neume-network/strategies/issues/170

**Review by TimDaub — CHANGES_REQUESTED** [review]
hard requirement is that all new functionality must be unit tested

**Comment by TimDaub** [comment]
> hard requirement is that all new functionality must be unit tested

Added issue for this too: https://github.com/neume-network/strategies/issues/174

**Comment by TimDaub** [comment]
- It's not a blocker for this PR but eventually we want to specify the graph as a json schema: https://github.com/neume-network/message-schema/issues/19

**Comment by il3ven** [comment]
I ran a small crawl using this PR. It ran successfully. I crawled 543 NFTs. The strategy was to limit the NFTs in web3subgraph-transformation. The crawl took 3m40s with a concurrency of 50. For IPFS I used `ipfs.io/ipfs` as my gateway.

Here is the crawl data if you are interested. [data.zip](https://github.com/neume-network/strategies/files/9101993/data.zip)

I think only unit tests are remaining to be added.

**Comment by il3ven** [comment]
Due to lack of reviewer, merging this branch to dev for now. So I can run the crawler without affecting master.

**Pull request merged**